### PR TITLE
Closes #6 | Add Loader for XCOPA

### LINKED ISSUE #6: Create dataset loader for XCOPA
Dataloader name: `xcopa/xcopa.py`
DataCatalogue: http://seacrowd.github.io/seacrowd-catalogue/card.html?xcopa

|     Dataset| xcopa |
|-------------|---|
| Description | XCOPA: A Multilingual Dataset for Causal Commonsense Reasoning The Cross-lingual Choice of Plausible Alternatives dataset is a benchmark to evaluate the ability of machine learning models to transfer commonsense reasoning across languages. The dataset is the translation and reannotation of the English COPA (Roemmele et al. 2011) and covers 11 languages from 11 families and several areas around the globe. The dataset is challenging as it requires both the command of world knowledge and the ability to generalise to new languages. |
| Subsets | th, id, vi |
| Languages | tha, vie, ind |
| Tasks | Commonsense Reasoning |
| License | Creative Commons Attribution 4.0 (cc-by-4.0) |
| Homepage | https://github.com/cambridgeltl/xcopa |
| HF URL | https://huggingface.co/datasets/xcopa |
| Paper URL | https://aclanthology.org/2020.emnlp-main.185/ |

### PR DESCRIPTION
**Pull request by FawwazMayda**
Closes #6 | Extend dataloader for XCOPA to `th` (Thai) and `vi` (Vietnamese)

### Checkbox
- [x] Confirm that this PR is linked to the dataset issue.
- [x] Create the dataloader script `seacrowd/sea_datasets/my_dataset/my_dataset.py` (please use only lowercase and underscore for dataset naming).
- [ ] Provide values for the `_CITATION`, `_DATASETNAME`, `_DESCRIPTION`, `_HOMEPAGE`, `_LICENSE`, `_URLs`, `_SUPPORTED_TASKS`, `_SOURCE_VERSION`, and `_SEACROWD_VERSION` variables.
- [x] Implement `_info()`, `_split_generators()` and `_generate_examples()` in dataloader script.
- [x] Make sure that the `BUILDER_CONFIGS` class attribute is a list with at least one `SEACrowdConfig` for the source schema and one for a seacrowd schema.
- [x] Confirm dataloader script works with `datasets.load_dataset` function.
- [x] Confirm that your dataloader script passes the test suite run with `python -m tests.test_seacrowd seacrowd/sea_datasets/<my_dataset>/<my_dataset>.py`.
- [ ] If my dataset is local, I have provided an output of the unit-tests in the PR (please copy paste). This is OPTIONAL for public datasets, as we can test these without access to the data files.

### Source
Screenshot 2024-01-21 at 23 13 07

### QA
Screenshot 2024-01-21 at 23 15 12

### Source
Screenshot 2024-01-21 at 23 16 40

### QA
Screenshot 2024-01-21 at 23 17 59

### Source
Screenshot 2024-01-21 at 23 18 57

### QA
Screenshot 2024-01-21 at 23 19 49

### CODE REVIEW

**Comment by sabilmakbar** [comment]
Hi @FawwazMayda, thanks for contributing on this dataloader!

May I know why Indonesian language is being skipped? Both the datacard and the source dataset has `id` as one of the supported languages, tho.

**Comment by FawwazMayda** [comment]
Currently it already has `id` in `seacrowd/xcopa` so because its already implemented that is why I left with implementing `th` and `vie`

Screenshot 2024-01-03 at 21 14 12

**Comment by sabilmakbar** [comment]
In that case, would you extend the script on `sea_datasets/xcopa` to cover `tha` and `vie` aswell? This shd be similar to this previously-merged PR, you may refer the implementation to here:
https://github.com/SEACrowd/seacrowd-datahub/pull/125

**Comment by FawwazMayda** [comment]
updated @sabilmakbar I added all those languages into single `xcopa.py`

**Review by holylovenia — CHANGES_REQUESTED** [review]
Hi @FawwazMayda, thanks a lot for your hard work!! 💪 Could I please request for these changes?

1. Re-arrange the field-value mapping for the `seacrowd_qa` schema a bit.

Given this `source` schema:
```
{
    'premise': premise,
    'choice1': choice1,
    'choice2': choice2,
    'question': reasoning_type,
    'label': label_id,
    'idx': id,
    'changed': changed_type
}
```

Let's use this `seacrowd_qa` structure:
```
{
    'id': id,
    'question_id': id,
    'document_id': id,
    'question': '',
    'type': 'multiple_choice',
    'choices': [
        choice1,
        choice2
    ],
    'context': premise,
    'answer': [
        choices[label_id]
    ],
    'meta': {
        'reasoning_type': reasoning_type,
        'is_changed': changed_type
    }
}
```

2. Use `_LANGUAGES = ["ind", "tha", "vie"]` instead of `_LANGUAGES = ["ind", "th", "vie"]`

3. Use `Licenses.CC_BY_4_0.value` instead of `"Unknown"` for `_LICENSE`.

**Review by holylovenia — APPROVED** [review]
LGTM! 👍 Thanks a lot, @FawwazMayda! Let's wait for @sabilmakbar's review now.

File: seacrowd/sea_datasets/xcopa/xcopa.py [hunk 1]
```diff
@@ -45,12 +45,20 @@
 
 _HOMEPAGE = "https://github.com/cambridgeltl/xcopa"
 
-_LICENSE = "Unknown"
+_LICENSE = "Licenses.CC_BY_4_0.value"
```
Comment (sabilmakbar) [inline seacrowd/sea_datasets/xcopa/xcopa.py:48]: this is an Enum object, not a string

```suggestion
_LICENSE = Licenses.CC_BY_4_0.value
```

**Review by sabilmakbar — COMMENTED** [review]
in [L11](https://github.com/SEACrowd/seacrowd-datahub/pull/286/files#diff-8054d53da69e3318748bfe9fc8843fe3a1e0e82613dbf86c356fabda660d71cdR11), add `License` object import to define the `_LICENSE` value correctly

```from seacrowd.utils.constants import Tasks, Licenses```

File: seacrowd/sea_datasets/xcopa/xcopa.py (hunk 1 above)
Comment (FawwazMayda) [reply seacrowd/sea_datasets/xcopa/xcopa.py]: added

**Comment by sabilmakbar** [comment]
Hi @FawwazMayda, can you try to run the formatter check and fix it accordingly?
I found some import & readability suggestions found like traling whitespace in a line, unused imports, fix reference of internal var in config creation (`l` to `lang`)

```sh
make check_file=seacrowd/sea_datasets/xcopa/xcopa.py
```

**Comment by FawwazMayda** [comment]
applied

**Comment by FawwazMayda** [comment]
updated @sabilmakbar

**Review by sabilmakbar — APPROVED** [review]
LGTM! Thanks @FawwazMayda and @holylovenia, merging this...

**Pull request merged**